### PR TITLE
use S-Express for SwankProtocolConversions

### PR DESCRIPTION
**Pull request by fommil**
Step 1 in #390

This should be completely backwards compatible with the emacs client, hence a lot of jiggery pokery in the new `SwankProtocolConversions.scala` file.

I've yet to run this on the emacs integration tests, it would be sooo good if we could get #777 up and running.

### CODE REVIEW

**Comment by fommil** [comment]
coverage decreasing because I'm removing covered code and replacing it with less code :smile:

File: src/main/scala/org/ensime/core/EnsimeEvent.scala [hunk 1]
```diff
@@ -1,31 +1,71 @@
...
 sealed trait GeneralSwankEvent extends EnsimeEvent
+sealed trait DebugEvent extends EnsimeEvent
 
+/** Generic background notification. */
 case class SendBackgroundMessageEvent(code: Int, detail: Option[String]) extends GeneralSwankEvent
+
+/** The presentation compiler is ready to accept requests. */
 case object AnalyzerReadyEvent extends GeneralSwankEvent
+
+/** The presentation compiler has finished analysing the entire project. */
```
Comment (rorygraves) [inline src/main/scala/org/ensime/core/EnsimeEvent.scala]: I like the doc comments

**Comment by aemoncannon** [comment]
Awesome stuff! Does this handle omitting nil / empty list values?

File: src/main/scala/org/ensime/sexp/formats/FamilyFormats.scala [hunk 2]
```diff
@@ -14,32 +14,37 @@ import org.ensime.sexp._
...
   case class TypeHint[T](hint: SexpSymbol)
   implicit def typehint[T: TypeTag]: TypeHint[T] =
-    TypeHint(SexpSymbol(":" + typeOf[T].dealias.toString.split("(\\.|\\$)").last))
+    TypeHint(SexpSymbol(":" + typeOf[T].dealias.toString.replaceAll("\\.type$", "").split("(\\.|\\$)").last))
+
+  // always serialises to Nil, and is differentiated by the TraitFormat
+  //
+  // this is not implicit because of a stray println in shapeless that
+  // makes this extremely frustrating to use.
+  // https://github.com/milessabin/shapeless/pull/259
```
Comment (rorygraves) [inline src/main/scala/org/ensime/sexp/formats/FamilyFormats.scala]: Are you going to have to remake this PR against Shapeless 2.1?

File: src/main/scala/org/ensime/sexp/formats/ProductFormats.scala [hunk 3]
```diff
@@ -94,3 +97,14 @@ trait ProductFormats extends LowPriorityProductFormats {
...
+
+/**
+ * By default, S-Express uses the same field names in the wire format
+ * as the code. But some protocols may prefer dashes with Scala code
+ * that uses camel case. This mix-in provides that behaviour.
+ */
+trait CamelCaseToDashes {
+  this: LowPriorityProductFormats =>
+  override def toWireName(field: String): String =
+    field.replaceAll("([A-Z])", "-$1").toLowerCase.replaceAll("^-", "")
```
Comment (rorygraves) [inline src/main/scala/org/ensime/sexp/formats/ProductFormats.scala]: neat

**Comment by fommil** [comment]
@aemoncannon yes, we can customise whatever we want but I think we're omitting entries with `nil` values for now. The problem is that any change like this involves manually updating so many tests :smile: The actual code change is about two lines.

File: src/main/scala/org/ensime/sexp/formats/FamilyFormats.scala (hunk 2 above)
Comment (fommil) [inline src/main/scala/org/ensime/sexp/formats/FamilyFormats.scala]: I tried upgrading to shapeless 2.1 but it broke in a non-trivial way that even miles didn't know the answer to. I think an upgrade will be easier to debug after I modularise (coming next!)

File: src/main/scala/org/ensime/core/EnsimeEvent.scala (hunk 1 above)
Comment (fommil) [inline src/main/scala/org/ensime/core/EnsimeEvent.scala]: yeah hopefully the API module will consist of only case classes with docs on them and therefore document all the various protocols.

File: src/main/scala/org/ensime/sexp/formats/FamilyFormats.scala (hunk 2 above)
Comment (rorygraves) [inline src/main/scala/org/ensime/sexp/formats/FamilyFormats.scala]: lol, fair enough.

File: src/test/scala/org/ensime/server/protocol/swank/SwankProtocolConversionsSpec.scala [hunk 4]
```diff
@@ -106,66 +106,62 @@ class SwankProtocolConversionsSpec extends FunSpec with Matchers {
...
         // toWF(value: NamedTypeMemberInfo)
-        assert(toWF(new NamedTypeMemberInfo("typeX", typeInfo, None, 'abcd)).toWireString === """(:name "typeX" :type (:name "type1" :type-id 7 :full-name "FOO.type1" :decl-as type1 :outer-type-id 8) :decl-as abcd)""")
+        assert(toWF(new NamedTypeMemberInfo("typeX", typeInfo, None, 'abcd)).toWireString === """(:name "typeX" :type (:arrow-type nil :name "type1" :type-id 7 :decl-as type1 :full-name "FOO.type1" :type-args nil :members nil :pos nil :outer-type-id 8) :pos nil :decl-as abcd)""")
 
         // toWF(value: EntityInfo)
         assert(toWF(entityInfo).toWireString === entityInfoStr)
 
         // toWF(value: TypeInfo)
-        assert(toWF(typeInfo).toWireString === """(:name "type1" :type-id 7 :full-name "FOO.type1" :decl-as type1 :outer-type-id 8)""")
+        assert(toWF(typeInfo).toWireString === """(:arrow-type nil :name "type1" :type-id 7 :decl-as type1 :full-name "FOO.type1" :type-args nil :members nil :pos nil :outer-type-id 8)""")
```
Comment (rorygraves) [inline src/test/scala/org/ensime/server/protocol/swank/SwankProtocolConversionsSpec.scala]: Does this mean the keys are always in a consistent order now?

**Comment by rorygraves** [comment]
build again

File: src/test/scala/org/ensime/server/protocol/swank/SwankProtocolConversionsSpec.scala (hunk 4 above)
Comment (fommil) [inline src/test/scala/org/ensime/server/protocol/swank/SwankProtocolConversionsSpec.scala]: yes, although that is **not** required on the input :smile:

**Comment by rorygraves** [comment]
This looks really good - anything else you want to do with it before its merged?

**Comment by fommil** [comment]
nope, everything else can go in as a separate PR

**Comment by aemoncannon** [comment]
+1 Maybe just run lisp integrations locally?

**Comment by fommil** [comment]
The emacs tests just hang when I run them

**Comment by rorygraves** [comment]
Thats interesting because jenkins is having the same issues right now:
e.g. http://fieldmarkeng.co.uk/job/Ensime-Emacs-CI-Build/46/

**Comment by rorygraves** [comment]
@fommil  The jenkins build failure is a transient around the long running debug tests - will fix that very shortly.

**Comment by rorygraves** [comment]
@fommil  Here you go - the tests are a lot quicker now: https://github.com/ensime/ensime-server/pull/831

**Comment by rorygraves** [comment]
(not sure why they failed randomly on the jenkins test)

**Comment by aemoncannon** [comment]
Here's the stacktrace from jenkins. Not sure how the test version could be nil...

Debugger entered!
  (error Invalid version string: '%s' nil)
  (version-to-list nil)
  (-take 2 (version-to-list ensime--test-scala-version))
  (mapconcat (quote int-to-string) (-take 2 (version-to-list ensime--test-scala-version)) .)
  (ensime--test-scala-major-version)
  (concat root-dir target/scala- (ensime--test-scala-major-version) /classes)
  (file-name-as-directory (concat root-dir target/scala- (ensime--test-scala-major-version) /classes))
  (let* ((root-dir (file-name-as-directory (make-temp-file ensime_test_proj_ t))) (cache-dir (file-name-as-directory (concat root-dir cache))) (src-dir (file-name-as-directory (concat root-dir src/main/scala))) (target-dir (file-name-as-directory (concat root-dir target/scala- (ensime--test-scala-major-version) /classes))) (test-target-dir (file-name-as-directory (concat root-dir test-target))) (config (append extra-config (list (quote :root-dir) root-dir (quote :cache-dir) cache-dir (quote :name) test (quote :scala-version) ensime--test-scala-version (quote :java-home) (getenv JAVA_HOME) (quote :subprojects) (list (list (quote :name) test (quote :module-name) test (quote :source-roots) (list src-dir) (quote :depends-on-modules) nil (quote :target) target-dir (quote :test-target) test-target-dir))))) (conf-file (ensime-create-file (concat root-dir .ensime) (format %S config)))) (mkdir src-dir t) (mkdir cache-dir t) (mkdir target-dir t) (mkdir test-target-dir t) (let* ((src-file-names (mapcar (function (lambda (f) (ensime-create-file (concat src-dir (plist-get f :name)) (plist-get f :contents)))) src-files))) (list :src-files src-file-names :root-dir root-dir :conf-file conf-file :src-dir src-dir :target target-dir)))
  (ensime-create-tmp-project ((:name test.scala :contents package pack.a
class A(value:/*1*/String){
})))

**Comment by fommil** [comment]
FWIW I've been using this today at work and it seems pretty stable

**Comment by fommil** [comment]
an exception in the logs

```
13:59:27.176 [default-akka.actor.default-dispatcher-2] INFO  o.e.s.protocol.swank.SwankProtocol - Received msg: (:swank-rpc (swank:symbol-at-point "OBFUSCATED" 1500) 8)
13:59:27.240 [default-akka.swank-dispatcher-14] ERROR o.e.s.protocol.swank.SwankProtocol - Exception whilst handling rpc expected SexpSymbol(empty)'s wrap to be SexpData
org.ensime.sexp.SerializationException: expected SexpSymbol(empty)'s wrap to be SexpData
	at org.ensime.sexp.formats.package$.serializationError(package.scala:6) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocolConversions$TraitFormatAlt.wrap(SwankProtocolConversions.scala:93) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocolConversions$SourcePositionFormat$.write(SwankProtocolConversions.scala:207) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocolConversions$SourcePositionFormat$.write(SwankProtocolConversions.scala:205) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.PimpedAny.toSexp(package.scala:13) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.OptionAltFormat$$anon$4.write(StandardFormats.scala:142) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.OptionAltFormat$$anon$4.write(StandardFormats.scala:140) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$$anon$1.write(ProductFormats.scala:66) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.PimpedAny.toSexp(package.scala:13) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocolConversions$TraitFormatAlt.wrap(SwankProtocolConversions.scala:88) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocolConversions$TypeInfoFormat$.write(SwankProtocolConversions.scala:333) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocolConversions$TypeInfoFormat$.write(SwankProtocolConversions.scala:328) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:32) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$HListFormat$$anon$4.write(ProductFormats.scala:31) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.LowPriorityProductFormats$$anon$1.write(ProductFormats.scala:66) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.PimpedAny.toSexp(package.scala:13) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.OptionAltFormat$$anon$4.write(StandardFormats.scala:142) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.formats.OptionAltFormat$$anon$4.write(StandardFormats.scala:140) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.sexp.PimpedAny.toSexp(package.scala:13) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocolConversions$.toWF(SwankProtocolConversions.scala:510) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocol.org$ensime$server$protocol$swank$SwankProtocol$$handleRPCRequest(SwankProtocol.scala:1126) ~[ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocol$$anonfun$handleEmacsRex$1.apply$mcV$sp(SwankProtocol.scala:122) [ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocol$$anonfun$handleEmacsRex$1.apply(SwankProtocol.scala:119) [ensime_2.11.jar:0.9.10-SNAPSHOT]
	at org.ensime.server.protocol.swank.SwankProtocol$$anonfun$handleEmacsRex$1.apply(SwankProtocol.scala:119) [ensime_2.11.jar:0.9.10-SNAPSHOT]
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.liftedTree1$1(Future.scala:24) [scala-library-2.11.5.jar:na]
	at scala.concurrent.impl.Future$PromiseCompletingRunnable.run(Future.scala:24) [scala-library-2.11.5.jar:na]
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41) [akka-actor_2.11-2.3.8.jar:na]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_51]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_51]
	at java.lang.Thread.run(Thread.java:744) [na:1.7.0_51]
```

**Comment by fommil** [comment]
it was a transient (likely timing in the pres compiler), I did the same query a moment later and it worked fine.

**Comment by fommil** [comment]
ok, that was a good stress test. I think the above issue is showing up in a few places... I need to fix it before this can be merged. I'll try to work out a test case and fix tonight (looks like it'll be pretty simple)

**Comment by fommil** [comment]
I think this transaction (which works in master) was breaking it

```
14:22:06.463 [default-akka.actor.default-dispatcher-8] INFO  o.e.s.protocol.swank.SwankProtocol - Received msg: (:swank-rpc (swank:symbol-at-point "OBFUSCATED" 1123) 7)
14:22:06.522 [Scala Presentation Compiler []] DEBUG o.e.core.RichPresentationCompiler - search: spray.routing.HttpServiceActor = Some(FqnSymbol(Some(8782429),file:///home/hallisam/.ivy2/cache/io.spray/spray-routing-shapeless2_2.11/bundles/spray-routing-shapeless2_2.11-1.3.2.jar,jar:file:///home/hallisam/.ivy2/cache/io.spray/spray-routing-shapeless2_2.11/bundles/spray-routing-shapeless2_2.11-1.3.2.jar!/spray/routing/HttpServiceActor.class,spray.routing.HttpServiceActor,None,None,Some(jar:file:///home/hallisam/.ivy2/cache/io.spray/spray-routing-shapeless2_2.11/srcs/spray-routing-shapeless2_2.11-1.3.2-sources.jar!/spray/routing/HttpService.scala),Some(96),None))
14:22:06.564 [default-akka.actor.default-dispatcher-8] INFO  o.e.s.protocol.swank.SwankProtocol - Writing: (:return (:ok (:name "spray.routing.HttpServiceActor" :local-name "HttpServiceActor" :type (:name "HttpServiceActor" :type-id 1 :full-name "spray.routing.HttpServiceActor" :decl-as class :pos t) :decl-pos (:file "/home/hallisam/starling/.ensime_cache/dep-src/source-jars/spray/routing/HttpService.scala" :offset 3241) :owner-type-id 2)) 7)
14:22:06.851 [default-akka.actor.default-dispatcher-8] INFO  o.e.s.protocol.swank.SwankProtocol - Received msg: (:swank-rpc (swank:symbol-designations "/home/hallisam/starling/.ensime_cache/dep-src/source-jars/spray/routing/HttpService.scala" 0 3325 (var val varField valField functionCall operator param class trait object package)) 8)
```

**Comment by aemoncannon** [comment]
Consistently failing here when I run locally - looks like an issue parsing the backtrace request?

Thread 'main' hit breakpoint at Test.scala : 7
Received test event: :debug-event with value (:type breakpoint :thread-id 1 :thread-name main :file /private/var/folders/4b/gj73872j2vs9wxtkrvg3dzb80000gn/T/ensime_test_proj_54241lpY/src/main/scala/test/Test.scala :line 7)
Handling test event: :debug-event
Received test event: :region-sem-highlighted with value nil
Got :region-sem-highlighted, expecting :debug-event. Ignoring event.
Received test event: :return-value with value (:ok (:file /private/var/folders/4b/gj73872j2vs9wxtkrvg3dzb80000gn/T/ensime_test_proj_54241lpY/src/main/scala/test/Test.scala :syms ((package 8 12) (param 38 42) (class 44 57) (val 69 70) (val 87 88) (val 105 106) (functionCall 120 127) (operator 134 135) (operator 130 131) (val 128 129) (val 132 133) (val 136 137))))
Got :return-value, expecting :debug-event. Ignoring event.
Synchronous RPC Aborted: Malformed swank:debug-backtrace call: (swank:debug-backtrace 1 0 -1)
Assertion failed at 'Test debugging scala project.': (ensime-test-assert-failed . Expected truth of trace.)
Error executing test: (ensime-test-interrupted . Test interrupted: (ensime-test-assert-failed . Expected truth of trace.).), moving to next.
Quitting ENSIME Scala
Server connection closed unexpectedly: connection broken by remote peer

Finished suite with status 1.

**Comment by fommil** [comment]
Interesting. I'm betting its a canonisation thing. We're canonising now whereas it probably wasn't before.

**Comment by fommil** [comment]
hmm, maybe this isn't canonisation stuff. I'd really like to be able to run the local emacs tests against this.

**Comment by fommil** [comment]
ok, simple fix! Turns out the "alt" trait format didn't handle case classes with no parameters because it was expecting a map of fields => values and the `SexpData` intentionally does not match on empty lists. Should be fixed now :smile: except for the emacs tests that aemon is seeing, although it may be related.

**Comment by fommil** [comment]
@aemoncannon this is where my local tests hang. Is there anything else I can give you to debug this?

```
Starting test: 'Test debugging scala project.'
ENSIME server starting...
Connecting to Swank on port 38033..
Received test event: :connected with value (:pid nil :implementation (:name ENSIME) :version 0.8.11)
Got :connected, expecting :compiler-ready. Ignoring event.
Connected to ENSIME speaking protocol 0.8.11, please wait while the project is loaded.
Received test event: :return-value with value (:ok (:pid nil :implementation (:name ENSIME) :version 0.8.11))
Got :return-value, expecting :compiler-ready. Ignoring event.
Received test event: :return-value with value (:ok t)
Got :return-value, expecting :compiler-ready. Ignoring event.
Initializing Analyzer. Please wait...
ENSIME ready. Let the hacking commence!
Received test event: :compiler-ready with value t
Handling test event: :compiler-ready
```

**Comment by aemoncannon** [comment]
I think I see the problem: thread ids used to have a manually conversion to String in SwankProtocolConversions, now they're being sent out as integral. Emacs just passes it through, and then it comes back to the server (in this case via debug-backtrace) and the StringAtom(threadId) pattern doesn't match.

**Comment by aemoncannon** [comment]
Maybe best to just declare them as string in the case class.

**Comment by aemoncannon** [comment]
@fommil (see above)

**Comment by fommil** [comment]
doh! There was **one thing** that both you and @rorygraves asked me about. `threadId` as `String` ONE THING. And what did I do? Yes, I changed it to `int`.... :facepunch:

**Comment by rorygraves** [comment]
lol

**Comment by aemoncannon** [comment]
: )  On the bright side the debug test is the last to run, so this is probably the only issue.

**Comment by rorygraves** [comment]
@fommil  - What have you done to my coverage! Do you know how hard it was to get it that far...  :anguished:

**Comment by rorygraves** [comment]
@aemoncannon That is definitely good news.

**Comment by fommil** [comment]
`git sith`d a fix. I'll try the local emacs tests.

**Comment by fommil** [comment]
the local tests still hang for me. Can you please try them again @aemoncannon ? Is it maybe something to do with the way we launch the JVM that we're debugging?

**Comment by fommil** [comment]
I can't see where we launch the debugged application with debugging enabled.

**Comment by aemoncannon** [comment]
@fommil gets launched by the jdi stuff internally. running them locally.

**Comment by aemoncannon** [comment]
@fommil I see
Assertion failed at 'Test debugging scala project.': (ensime-test-assert-failed . Expected frame to equal (:index 0 :locals ((:index 0 :name\
 "args" :summary "Array[]" :type-name "java.lang.String[]") (:index 1 :name "a" :summary "\"cat\"" :type-name "java.lang.String") (:index 2 \
:name "b" :summary "\"dog\"" :type-name "java.lang.String") (:index 3 :name "c" :summary "\"bird\"" :type-name "java.lang.String")) :num-arg\
s 1 :class-name "test.Test$" :method-name "main" :pc-location (:file "/private/var/folders/4b/gj73872j2vs9wxtkrvg3dzb80000gn/T/ensime_test_p\
roj_55245R3z/src/main/scala/test/Test.scala" :line 7) :this-object-id "NA"), but was (:index 0 :locals ((:index 0 :name "args" :type-name "j\
ava.lang.String[]" :summary "Array[]") (:index 1 :name "a" :type-name "java.lang.String" :summary "\"cat\"") (:index 2 :name "b" :type-name \
"java.lang.String" :summary "\"dog\"") (:index 3 :name "c" :type-name "java.lang.String" :summary "\"bird\"")) :num-args 1 :class-name "test\
.Test$" :method-name "main" :pc-location (:file "/private/var/folders/4b/gj73872j2vs9wxtkrvg3dzb80000gn/T/ensime_test_proj_55245R3z/src/main\
/scala/test/Test.scala" :line 7) :this-object-id "NA").)

**Comment by aemoncannon** [comment]
@fommil  i.e. a test that depends on the order of elements of a plist. will the new order be stable enough that i can just paste this into the test to fix?

**Comment by aemoncannon** [comment]
@fommil actually might be easier to just change the order in the case class if that suits you

**Comment by fommil** [comment]
Thanks! Changing to case  class should be easy.

**Comment by fommil** [comment]
That problem is still not fixed

```
Exception whilst handling rpc expected SexpSymbol(empty)'s wrap to be SexpData, was SexpSymbol(t)
```

I'll investigate and roll up with the changes to the case class.

**Comment by fommil** [comment]
I'm able to reproduce the stack trace I was seeing in a unit test, and unfortunately the fix is a change in the protocol. Luckily I think its in an area the client never uses --- a single `EmptySourcePosition` not part of a larger packet.

I'm still unable to run the emacs tests but I'm investigating installing an emacs23 to test against.

**Comment by fommil** [comment]
no emacs23 for debian jessie

**Comment by fommil** [comment]
this should fix it

**Comment by fommil** [comment]
found the problem! the ensime-tests are using `JAVA_HOME` and they should really be using `JDK_HOME` followed by `JAVA_HOME` followed by the OS X hack. My `JAVA_HOME` points to the JRE intentionally to detect these problems :stuck_out_tongue: (reconsidering that choice... although it is good to catch it early)

**Comment by fommil** [comment]
You'll have to trust me, but green lights in the emacs tests :smile:

Now that they work locally, I'm pretty sure I can get them working in travis.

**Comment by aemoncannon** [comment]
Nice

**Pull request merged**